_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+[1] "conicalOnR3" "cylindricalOnR3" "dihedralOnR3" "flip2dOnR2" "flipRot2dOnR2" "flipRot3dOnR3"
+[7] "fullCylindricalOnR3" "fullIcoOnR3" "fullOctaOnR3" "icoOnR3" "invOnR3" "mirOnR3 "octaOnR3"
+[14] "rot2dOnR2" "rot2dOnR3" "rot3dOnR3" "trivialOnR2" "trivialOnR3"    
+```
+
+The methods I've listed instantiate a `gspace`. If you look closely, you see that they're all composed of two strings, joined by "On". In all instances, the second part is either `R2` or `R3`. These two are the available base spaces -- $\mathbb{R}^2$ and $\mathbb{R}^3$ -- an input signal can live in. Signals can, thus, be images, made up of pixels, or three-dimensional volumes, composed of voxels. The first part refers to the group you'd like to use. Choosing a group means choosing the symmetries to be respected. For example, `rot2dOnR2()` implies equivariance as to rotations, `flip2dOnR2()` guarantees the same for mirroring actions, and `flipRot2dOnR2()` subsumes both.
+
+Let's define such a `gspace`. Here we ask for rotation equivariance on the Euclidean plane, making use of the same cyclic group -- $C_4$ -- we developed in our [from-scratch implementation](https://blogs.rstudio.com/ai/posts/2023-03-27-group-equivariant-cnn-2/):
+
+```{r}
+r2_act <- gspaces$rot2dOnR2(N = 4L)
+r2_act$fibergroup
+```
+
+In this post, I'll stay with that setup, but we could as well pick another rotation angle -- `N = 8`, say, resulting in eight equivariant positions separated by forty-five degrees. Alternatively, we might want *any* rotated position to be accounted for. The group to request then would be [SO(2)](https://en.wikipedia.org/wiki/Orthogonal_group#Special_orthogonal_group), called the *special orthogonal group,* of continuous, distance- and orientation-preserving transformations on the Euclidean plane:
+
+```{r}
+(gspaces$rot2dOnR2(N = -1L))$fibergroup
+```
+
+```         
+SO(2)
+```
+
+Going back to $C_4$, let's investigate its [*representations*](https://en.wikipedia.org/wiki/Group_representation):
+
+```{r}
+r2_act$representations
+
+```
+
+```         
+$irrep_0
+C4|[irrep_0]:1
+
+$irrep_1
+C4|[irrep_1]:2
+
+$irrep_2
+C4|[irrep_2]:1
+
+$regular
+C4|[regular]:4
+```
+
+A representation, in our current context *and* very roughly speaking, is a way to encode a group action as a matrix, meeting certain conditions. In `escnn`, representations are central, and we'll see how in the next section.
+
+First, let's inspect the above output. Four representations are available, three of which share an important property: they're all [irreducible](https://en.wikipedia.org/wiki/Irreducible_representation). On $C_4$, any non-irreducible representation [can be decomposed](https://en.wikipedia.org/wiki/Maschke%27s_theorem) into into irreducible ones. These irreducible representations are what `escnn` works with internally. Of those three, the most interesting one is the second. To see its action, we need to choose a group element. How about counterclockwise rotation by ninety degrees:
```

Comment:
Again, really good stuff—very clear even to someone who isn't familiar with the vocabulary.

Reply:
thanks!!